backend-lib/src/main/rust/lib.rs

```diff
@@ -2416,13 +2453,43 @@ pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_model_TorClient_freeTorRu
     _: JClass<'local>,
     ptr: jlong,
 ) {
-    let ptr = (ptr as usize) as *mut crate::tor::TorRuntime;
+    let ptr = std::ptr::with_exposed_provenance_mut::<crate::tor::TorRuntime>(ptr as usize);
     if !ptr.is_null() {
         let s = unsafe { Box::from_raw(ptr) };
         drop(s);
     }
 }
 
+/// Returns a new isolated `TorClient` handle.
+///
+/// The two `TorClient`s will share internal state and configuration, but their streams
+/// will never share circuits with one another.
+///
+/// Use this method when you want separate parts of your program to each have a
+/// `TorClient` handle, but where you don't want their activities to be linkable to one
+/// another over the Tor network.
+///
+/// Calling this method is usually preferable to creating a completely separate
+/// `TorClient` instance, since it can share its internals with the existing `TorClient`.
+#[unsafe(no_mangle)]
+pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_model_TorClient_isolatedClient<'local>(
+    mut env: JNIEnv<'local>,
+    _: JClass<'local>,
+    tor_runtime: jlong,
+) -> jlong {
+    let res = panic::catch_unwind(|| {
+        let tor_runtime =
+            ptr::with_exposed_provenance_mut::<crate::tor::TorRuntime>(tor_runtime as usize);
+        let tor_runtime =
+            unsafe { tor_runtime.as_mut() }.ok_or_else(|| anyhow!("A Tor runtime is required"))?;
+
+        let isolated_client = tor_runtime.isolated_client();
+
+        Ok(Box::into_raw(Box::new(isolated_client)).expose_provenance() as jlong)
+    });
+    unwrap_exc_or(&mut env, res, -1)
```

Comment:
Rust 1.81.0 started [aborting on uncaught panics in `extern "C"` functions](https://blog.rust-lang.org/2024/09/05/Rust-1.81.0.html#abort-on-uncaught-panics-in-extern-c-functions), which fixed the soundness hole that I think originally caused us to *have* to use `panic::catch_unwind` for safety. We should reconsider whether we want to continue to use this pattern, especially in cases like this one where a panic could leave a mutable object (like `tor_runtime`) in an undefined state.

Non-blocking; filed #1692.